numbat/src/parser.rs

```diff
@@ -219,31 +220,46 @@ struct Parser<'a> {
     tokens: &'a [Token],
     current: usize,
     decorator_stack: Vec<Decorator>,
+    stop_on_error: bool,
 }
 
 impl<'a> Parser<'a> {
-    pub(crate) fn new(tokens: &'a [Token]) -> Self {
+    pub(crate) fn new(tokens: &'a [Token], stop_on_error: bool) -> Self {
         Parser {
             tokens,
             current: 0,
             decorator_stack: vec![],
+            stop_on_error,
         }
     }
 
     fn skip_empty_lines(&mut self) {
         while self.match_exact(TokenKind::Newline).is_some() {}
     }
 
-    fn parse(&mut self) -> Result<Vec<Statement>> {
+    /// Parse a token stream.
+    /// If an error is encountered and `stop_on_error` is set to false, the parser
+    /// will try to recover from the error and parse as many statements as possible
+    /// while stacking all the errors in a `Vec`. At the end, it returns the complete
+    /// list of statements parsed + the list of errors accumulated.
+    fn parse(&mut self) -> ParseResult {
         let mut statements = vec![];
+        let mut errors = vec![];
 
         self.skip_empty_lines();
 
         while !self.is_at_end() {
             match self.statement() {
                 Ok(statement) => statements.push(statement),
+                Err(e) if self.stop_on_error => {
+                    return Err((statements, vec![e]));
+                }
                 Err(e) => {
-                    return Err(e);
+                    errors.push(e);
+                    // skip all the tokens until we encounter a newline.
+                    while !matches!(self.peek().kind, TokenKind::Newline | TokenKind::Eof) {
+                        self.advance()
+                    }
```

Comment:
Very cool :star_struck:. I read about error-recovery in parsers, but somehow I was afraid it would be much more complicated then this (well, I guess it can be, but I didn't expect we could start with something like this).

Reply:
Huuum yes, I didn't dive deep into it, but I think it can get really hard, especially in languages that do not have an « end of statements » symbol.
But since `numbat` statements always end with a `\n` it helps (and when they don't, like for function or let statements, it's not a big issue either).